interp/interp.go

```diff
@@ -146,18 +152,146 @@ func (pc PiecewiseConstant) Predict(x float64) float64 {
 	if i < 0 {
 		return pc.ys[0]
 	}
-	// i < len(pci.xs)
 	if x == pc.xs[i] {
 		return pc.ys[i]
 	}
 	n := len(pc.xs)
 	if i == n-1 {
-		// x > pci.xs[i]
 		return pc.ys[n-1]
 	}
 	return pc.ys[i+1]
 }
 
+// PiecewiseCubic is a piecewise cubic 1-dimensional interpolator with
+// continuous value and first derivative.
+type PiecewiseCubic struct {
+	// Interpolated X values.
+	xs []float64
+
+	// Coefficients of interpolating cubic polynomials, with
+	// len(xs) - 1 rows and 4 columns. The interpolated value
+	// for xs[i] <= x < xs[i + 1] is defined as
+	//   sum_{k = 0}^3 coeffs.At(i, k) * (x - xs[i])^k
+	// To guarantee left-continuity, coeffs.At(i, 0) == ys[i].
+	coeffs *mat.Dense
+
+	// Last interpolated Y value, corresponding to xs[len(xs) - 1].
+	lastY float64
+}
+
+// Predict returns the interpolation value at x.
+func (pc PiecewiseCubic) Predict(x float64) float64 {
+	i := findSegment(pc.xs, x)
+	if i < 0 {
+		return pc.coeffs.At(0, 0)
+	}
+	m := len(pc.xs) - 1
+	if x == pc.xs[i] {
+		if i < m {
+			return pc.coeffs.At(i, 0)
+		}
+		return pc.lastY
+	}
+	if i == m {
+		return pc.lastY
+	}
+	dx := x - pc.xs[i]
+	a := pc.coeffs.RawRowView(i)
+	return ((a[3]*dx+a[2])*dx+a[1])*dx + a[0]
+}
+
+// fitWithDerivatives fits a piecewise cubic predictor to (X, Y, dY/dX) value
+// triples provided as three slices.
+// It panicks if len(xs) < 2, elements of xs are not strictly increasing,
+// len(xs) != len(ys) or len(xs) != len(dydxs).
+func (pc *PiecewiseCubic) fitWithDerivatives(xs, ys, dydxs []float64) {
```

Comment:
Unexporting this function makes this type unusable because there's no way to fit it. I would export this function and embed `PiecewiseCubic` as an unexported field in `Akima`.

Reply:
Right, that's the pattern I wanted all the time.

Reply:
Oops.